core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -896,6 +913,193 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord]) extends PSBTMap {
+      }
+
+      val records = utxos ++ recordsToAdd ++ unknowns
+      InputPSBTMap(records)
+    }
+
+    /** Turns the required PartialSignatures into a ScriptSignature and calls wipeAndAdd
+      * @return None if the requirement is not met
+      */
+    def collectSigs(
+        required: Int,
+        constructScriptSig: Seq[(ECDigitalSignature, ECPublicKey)] => ScriptSignature): Option[
+      InputPSBTMap] = {
+      val sigs = getRecords[PartialSignature](PartialSignatureKeyId)
+      if (sigs.length != required) {
+        None
+      } else {
+        val scriptSig = constructScriptSig(
+          sigs.map(sig => (sig.signature, sig.pubKey)))
+
+        val newInputMap = wipeAndAdd(scriptSig)
+
+        Some(newInputMap)
+      }
+    }
+
+    spkToSatisfy match {
+      case _: P2SHScriptPubKey =>
+        val redeemScriptOpt =
+          getRecords[RedeemScript](RedeemScriptKeyId).headOption
+        redeemScriptOpt.flatMap {
+          case InputPSBTRecord.RedeemScript(redeemScript) =>
+            finalize(redeemScript).map { inputMap =>
+              val nestedScriptSig = inputMap
+                .getRecords[FinalizedScriptSig](FinalizedScriptSigKeyId)
+                .head
+              val witnessOpt = inputMap
+                .getRecords[FinalizedScriptWitness](FinalizedScriptWitnessKeyId)
+                .headOption
+
+              val scriptSig =
+                P2SHScriptSignature(nestedScriptSig.scriptSig, redeemScript)
+              wipeAndAdd(scriptSig, witnessOpt.map(_.scriptWitness))
+            }
+        }
+      case _: P2WSHWitnessSPKV0 =>
+        val redeemScriptOpt =
+          getRecords[WitnessScript](WitnessScriptKeyId).headOption
+        redeemScriptOpt.flatMap {
+          case WitnessScript(redeemScript) =>
+            finalize(redeemScript).map { inputMap =>
+              val nestedScriptSig = inputMap
+                .getRecords[FinalizedScriptSig](FinalizedScriptSigKeyId)
+                .head
+              val scriptSig = EmptyScriptSignature
+              wipeAndAdd(
+                scriptSig,
+                Some(P2WSHWitnessV0(redeemScript, nestedScriptSig.scriptSig)))
+            }
+        }
+      case _: P2WPKHWitnessSPKV0 =>
+        val sigOpt =
+          getRecords[PartialSignature](PartialSignatureKeyId).headOption
+        sigOpt.map { sig =>
+          val witness = P2WPKHWitnessV0(sig.pubKey, sig.signature)
+          val scriptSig = EmptyScriptSignature
+          wipeAndAdd(scriptSig, Some(witness))
+        }
+      case conditional: ConditionalScriptPubKey =>
+        val builder =
+          Vector.newBuilder[(
+              ConditionalPath,
+              Vector[Sha256Hash160Digest],
+              RawScriptPubKey)]
+
+        /** Traverses the ConditionalScriptPubKey tree for leaves and adds them to builder */
+        def addLeaves(rawSPK: RawScriptPubKey, path: Vector[Boolean]): Unit = {
```

Comment:
can this be marked as `tailrec`?

Reply:
sadly no, since there are two calls made to `addLeaves` in the `ConditionalScriptPubKey` case. The good news is that people don't do giant nesting of `ConditionalScriptPubKey`s and I doubt it is even possible to do so to such an extent that this recursion gets out of hand